pkg/services/mssqldr/common_provision.go

```diff
@@ -68,3 +71,185 @@ func validateServer(
 	}
 	return *result.FullyQualifiedDomainName, nil
 }
+
+func validateDatabase(
+	ctx context.Context,
+	databasesClient *sqlSDK.DatabasesClient,
+	resourceGroup string,
+	serverName string,
+	databaseName string,
+) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	_, err := databasesClient.Get(
+		ctx,
+		resourceGroup,
+		serverName,
+		databaseName,
+		"",
+	)
+	if err != nil {
+		return fmt.Errorf("error getting the sql database: %s", err)
+	}
+	// TODO: add the plan as param and validate?
+	return nil
+}
+
+func validateFailoverGroup(
+	ctx context.Context,
+	failoverGroupsClient *sqlSDK.FailoverGroupsClient,
+	resourceGroup string,
+	priServerName string,
+	secServerName string,
+	databaseName string,
+	failoverGroupName string,
+) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	result, err := failoverGroupsClient.Get(
+		ctx,
+		resourceGroup,
+		priServerName,
+		failoverGroupName,
+	)
+	if err != nil {
+		return fmt.Errorf("error getting the failover group: %s", err)
+	}
+	if len(*result.PartnerServers) > 1 {
+		return fmt.Errorf("error unexpected more than one partner server")
+	}
+	partnerServerRole := (*result.PartnerServers)[0].ReplicationRole
+	if partnerServerRole != sqlSDK.Secondary {
+		return fmt.Errorf("error partner server is not " +
+			"the secondary role in the failover group")
+	}
+	partnerServerID := *(*result.PartnerServers)[0].ID
+	serverNameReStr := regexp.MustCompile("^.*servers/([-a-zA-Z0-9]+)$")
+	partnerServerName := serverNameReStr.ReplaceAllString(
+		partnerServerID,
+		"$1",
+	)
+	if partnerServerName != secServerName {
+		return fmt.Errorf("error partner server is not the one specified")
+	}
+
+	if len(*result.Databases) > 1 {
+		return fmt.Errorf("error unexpected more than one database " +
+			"in the failover group")
+	}
+	actualDatabaseID := (*result.Databases)[0]
+	databaseNameReStr := regexp.MustCompile("^.*databases/([-a-zA-Z0-9]+)$")
+	actualDatabaseName := databaseNameReStr.ReplaceAllString(
+		actualDatabaseID,
+		"$1",
+	)
+	if actualDatabaseName != databaseName {
+		return fmt.Errorf("error unexpected database in the failover group")
+	}
+	return nil
+}
+
+func buildDatabaseGoTemplateParameters(
+	databaseName string,
+	pp service.ProvisioningParameters,
+	pd planDetails,
+) (map[string]interface{}, error) {
+	td, err := pd.getTierProvisionParameters(pp)
+	if err != nil {
+		return nil, err
+	}
+	p := map[string]interface{}{}
+	p["databaseName"] = databaseName
+	for key, value := range td {
+		p[key] = value
+	}
+	return p, nil
+}
+
+func deployDatabaseARMTemplate(
+	armDeployer *arm.Deployer,
```

Comment:
Why `armDeployer` is a pointer? Seems there is no change to the object. Call by value is enough...?

Reply:
Yeah, no change. When I wrote these, I thought `arm.Deployer` is big enough to pass by pointer instead of value. Does it make sense?